tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Losses.java

```diff
@@ -0,0 +1,685 @@
+        if (output.op().numOutputs() != 1)
+          throw new IllegalArgumentException("output can only have 1 output");
+        output = output.op().output(0);
+        return tf.nn.sigmoidCrossEntropyWithLogits(target, output);
+      }
+    }
+    DataType<T> dataType = output.asOutput().dataType();
+    Operand<T> one = tf.dtypes.cast(tf.constant(1), dataType);
+    Operand<T> epsilonConst = tf.dtypes.cast(tf.constant(EPSILON), dataType);
+    Operand<T> oneMinusEpsilonConst = tf.math.sub(one, epsilonConst);
+    output = tf.clipByValue(output, epsilonConst, oneMinusEpsilonConst);
+
+    // Compute cross entropy from probabilities.
+    Operand<T> bce = tf.math.mul(target, tf.math.log(tf.math.add(output, epsilonConst)));
+    bce =
+        tf.math.add(
+            bce,
+            tf.math.mul(
+                tf.math.sub(one, target),
+                tf.math.log(tf.math.add(tf.math.sub(one, output), epsilonConst))));
+    return tf.math.neg(bce);
+  }
+
+  /**
+   * Computes the categorical crossentropy loss between labels and predictions.
+   *
+   * @param tf the TensorFlow Ops
+   * @param labels true targets
+   * @param predictions the predictions
+   * @param fromLogits Whether to interpret predictions as a tensor of logit values
+   * @param labelSmoothing Float in [0, 1]. When 0, no smoothing occurs. When &gt; 0, compute the
+   *     loss between the predicted labels and a smoothed version of the true labels, where the
+   *     smoothing squeezes the labels towards 0.5. Larger values of labelSmoothing correspond to
+   *     heavier smoothing.
+   * @param axis the
+   * @param <T> the data type of the predictions and labels
+   * @return the categorical crossentropy loss.
+   */
+  public static <T extends TNumber, U extends TNumber> Operand<T> categoricalCrossentropy(
+      Ops tf,
+      Operand<U> labels,
+      Operand<T> predictions,
+      boolean fromLogits,
+      float labelSmoothing,
+      int axis) {
+    DataType<T> dataType = predictions.asOutput().dataType();
+    Operand<T> tLabels = tf.dtypes.cast(labels, dataType);
+    LossTuple<T> ops = LossesImpl.squeezeOrExpandDimensions(tf, tLabels, predictions, null);
+    predictions = ops.getTarget();
+    tLabels = ops.getLabels();
+
+    if (labelSmoothing != 0.0f) {
+      tLabels = smoothLabelsCatX(tf, tLabels, labelSmoothing);
+    }
+    if (fromLogits) {
+      return tf.nn.softmaxCrossEntropyWithLogits(tLabels, predictions, -1);
+    }
+    if (!(predictions instanceof Variable) && (!tf.scope().env().isEager())) {
+      // TODO output = backtrackIdentity(output); doesn't seem to work with Java version.
+      if (predictions.op().type().equals("Softmax")) {
+        if (predictions.op().numOutputs() != 1)
+          throw new IllegalArgumentException("output can only have 1 output");
+        predictions = predictions.op().output(0);
+        return tf.nn.softmaxCrossEntropyWithLogits(tLabels, predictions, -1);
```

Comment:
This was a TODO and the output() what just  a holder until I figured out how to get the inputs to the
`SoftMax` operation. If you don't have a suggestion, maybe remove this logic for now.

Here is the logic from Python.

```
    # When softmax activation function is used for output operation, we
    # use logits from the softmax function directly to compute loss in order
    # to prevent collapsing zero when training.
    # See b/117284466
    assert len(output.op.inputs) == 1
    output = output.op.inputs[0]
    return nn.softmax_cross_entropy_with_logits_v2(
        labels=target, logits=output, axis=axis)
```

Reply:
I commented out this code and marked it as a TODO.